operations/operations.py

```diff
@@ -0,0 +1,109 @@
+from functools import partial
+
+import numpy as np
+import scipp as sc
+
+
+def mask_from_adj_pixels(mask, bank_width):
+    """
+    Checks if the adjacent pixels (in 8 directions) are masked to remove
+    any noisy pixels which are erroneously masked or unmasked compared to
+    it's neighbours
+
+    If all adj. pixels are then the pixel considered is set to True
+    If no adj. pixels are then the pixel considered is set to False
+    If surrounding pixels have a mix of True/False the val is left as-is
+
+    This function handles border pixels as if they aren't there. So that
+    the following happens:
+    ------------------------
+    |F|T|     ->      |T|T|
+    |T|T|             |T|T|
+    -----------------------
+
+    Parameters
+    ----------
+    mask: Existing mask with some positions masked
+    bank_width: The width of each bank to reshape into 2D
+
+    Returns
+    -------
+    mask: Mask copy after completing the op. described above
+
+    """
+
+    def shift(var, dim, forward, out_of_bounds):
+        fill = var[dim, 0:1].copy()
+        fill.values = np.full_like(fill.values, out_of_bounds)
+        if forward:
+            return sc.concatenate(fill, var[dim, :-1], dim)
+        else:
+            return sc.concatenate(var[dim, 1:], fill, dim)
+
+    mask = mask.copy()
+    mask = sc.reshape(mask, dims=["y", "x"], shape=(bank_width, bank_width))
+
+    def make_flip(fill):
+        flip = sc.Variable(dims=['neighbor', 'y', 'x'],
+                           shape=[8, ] + mask.shape, dtype=sc.dtype.bool)
+        flip['neighbor', 0] = shift(mask, "x", True, fill)
+        flip['neighbor', 1] = shift(mask, "x", False, fill)
+        flip['neighbor', 2] = shift(mask, "y", True, fill)
+        flip['neighbor', 3] = shift(mask, "y", False, fill)
+        flip['neighbor', 4:6] = shift(flip['neighbor', 0:2], "y", True, fill)
+        flip['neighbor', 6:8] = shift(flip['neighbor', 0:2], "y", False, fill)
+        return flip
+
+    # mask if all neighbors masked
+    mask = mask | sc.all(make_flip(True), 'neighbor')
+    # unmask if no neighbor masked
+    mask = mask & sc.any(make_flip(False), 'neighbor')
+
+    # Flatten using numpy, avoids #1192
+    return sc.Variable(["spectrum"], values=mask.values.ravel())
+
+
+def _calc_adj_spectra(center_spec_num, bank_width, num_spectra):
```

Comment:
Remind me why we still do this pixel-by pixel with the hack of using a thread pool?

Reply:
Originally I replaced this with a median filter, however there is some nuance to the original implementation this misses which results in incorrect results.

After spending multiple days trying to nail down what is wrong with say using a median filter in SciPy or how to nicely do a 2D kernel in Numpy without using complex stride tricks I reverted back to this and left it as a TODO for another developer to come back to in a future PR as other changes weren't getting done. 

I've left this hack in with plenty of TODOs in the main notebook to replace it whilst I worked on other feedback

Reply:
I could drop the threadpool, but I'd rather do it properly that fix a hack